🐛 fix: expert의 캐리어 업데이트 오류 수정

SongMihyun opened the pull request:
### 📌 관련 이슈
- 이슈 번호: # (연관된 이슈 번호를 적어주세요)

### 📝 작업 내용
-[x] expert 의 경력 업데이트시 오류 수정 / 기존경력 모두 삭제후 새로 경력 생성하는 로직으로 변경
- [ ] 변경 사항 설명 2
- [ ] 기타 필요한 설명

### 📋 체크리스트
- [x] 코드에 오류가 없는지 확인했습니다.
- [ ] 문서가 업데이트되었습니다.
- [ ] 테스트를 통과했습니다.

### 🔗 참고 자료
- 추가로 참고할 자료가 있다면 링크를 추가하세요.

### 아래 # 뒤에 PR머지와 함께 Close할 이슈번호를 입력해주세요.
closes #

Commit:
🐛 fix: expert의 캐리어 업데이트 오류 수정

- expert 의 경력 업데이트시 오류 수정 / 기존경력 모두 삭제후 새로 경력 생성하는 로직으로 변경

Commit:
test.sh

Meoyoug commented on src/expert/seriailzers.py:
```diff
@@ -113,6 +111,23 @@ def create(self, validated_data):
+            Career.objects.filter(expert=instance).delete()
+
+            # 새로운 Career 객체 생성
+            Career.objects.bulk_create([Career(expert=instance, **career) for career in careers_data])
```
미현님 여기 들여쓰기 이상해요..

Meoyoug reviewed (commented):
들여쓰기 틀린 부분만 고쳐서 다시 올려주세요

Commit:
♻️ refactor: 소셜로그인,로그아웃 시리얼라이저,뷰 리팩토링

- 리프레시 토큰 401 에러 수정
- 토큰 재발급 요청시 리프레시토큰을 헤더에 포함해서 받는걸로 코드변경

Meoyoug commented on src/users/oauth_serializers.py:
```diff
@@ -0,0 +1,217 @@
+from users.models import User
+
+
+class AllowExpiredTokenPermission(BasePermission):
```
이거는 common 에 permissions에 생성해주세요

SongMihyun replied:
옮기도록할게용

Meoyoug commented on src/users/oauth_serializers.py:
```diff
@@ -0,0 +1,217 @@
+    profile_image = serializers.URLField(required=False, allow_blank=True)
+    phone_number = serializers.CharField(max_length=20, required=False, allow_blank=True)
+
+    def validate_email(self, value):
```
검증을 할거면 이메일 형식인지를 검증해야할 것 같아요 기본적으로 해당 필드의 값이 존재하는 지 확인하는 작업은 시리얼라이저에서 required 속성에 따라 해줍니다.

Meoyoug commented on src/users/oauth_serializers.py:
```diff
@@ -0,0 +1,217 @@
+        전화번호 유효성 검증 및 표준화
+        - 전화번호는 항상 010-0000-0000 형태로 변환.
+        """
+        if not value:
```
이거는 전화번호를 제공하지 않는 경우도 있어서 None으로 하신건가요?

Meoyoug commented on src/users/oauth_serializers.py:
코드 맥락의 의미를 모르겠어요..
굳이 저 모든걸 시리얼라이저에서 수행할 필요가 있었나 싶네요
모델 시리얼라이저를 사용해서 User 객체에 대해서 액세스 토큰을 통해서 정보를 가져온 후에 save()에서 유저 객체를 저장하려고 했던걸로 보이는데 맞나요..?

그리고 계속해서 OutStandingToken에 직접 접근해서 토큰을 가져와서 BlackList에 추가한 후에 해당 토큰을 delete 하려고 하시는데
쿼리 횟수도 늘고 좋은 접근은 아닌 것 같아요

어차피 인증 미들웨어에서 JWT 토큰이 만료되면 401 에러를 반환해서 토큰에 대한 작업처리는 안해도 될거같아요

원래 SimpleJWT에 있는 Token Refresh View를 활용하는게 더 좋아보이네요,,

SongMihyun replied:
넹 알겠습니당~

Commit:
🐛 fix: profile_image url형태로 가공 /
♻️ refactor:

-[x]로그인시 전화번호로 사용자검증하던로직 삭제 - 이메일로만 검증//
-[x]프로플사진 앞에 붙어있던 /media/값 삭제 및 url 주소로 제대로 나오게 변경

Commit:
test.sh

Meoyoug commented on src/users/oauth_serializers.py:
```diff
@@ -0,0 +1,235 @@
+        header = self.get_header(request)  # Authorization 헤더 가져오기
+        raw_token = self.get_raw_token(header)  # 토큰 값 추출
+
+        # 특정 경로에서는 Authorization 헤더가 없어도 동작 가능하도록 설정
```
이걸 퍼미션으로 동작하게 이미 만들어놨는데 왜 JWTAuthentication 을 오버라이딩해서 또 검증하나요..?

SongMihyun replied:
하나만 했더니 중간에 한번더 검증해서 막히더라구요.... 어째야하나 하다가.... 둘다하면 된다고 하길래 해본거였습니다. ;

Meoyoug commented on src/users/oauth_serializers.py:
```diff
@@ -0,0 +1,235 @@
+                raise UnauthorizedException(detail="유효하지 않은 액세스 토큰입니다.", code="INVALID_ACCESS_TOKEN")
+
+
+class AccessTokenSerializer(serializers.Serializer):
```
시리얼라이저는 필드의 유효성 검증이 주 기능인데 토큰생성에 필요한거면굳이 왜 이걸 시리얼라이저라고 하셧는지 모르겠네요 유틸성이 더 커보입니다.

SongMihyun replied:
네엥

Meoyoug commented on src/users/oauth_serializers.py:
```diff
@@ -0,0 +1,235 @@
+        if not value:
+            raise BadRequestException("리프레시 토큰이 필요합니다.", code="MISSING_REFRESH_TOKEN")
+
+        if len(value) > 512:
```
필드에 max_length 속성을 지정했으면 validate메서드에서 따로 검증안해도 수행됩니다.

Meoyoug commented on src/users/oauth_serializers.py:
```diff
@@ -0,0 +1,235 @@
+
+        try:
+            token = RefreshToken(value)
+        except Exception as e:
```
```
 try:
            token = RefreshToken(value)
```
이 부분에서 토큰 검증할 때 발생할 수 있는 예외들에 대해서 포괄적으로 한번에 처리하지말고 어떤 예외가 발생했는지 알 수 있도록 예외처리 세분화 해주세요

Meoyoug commented on src/users/oauth_serializers.py:
```diff
@@ -0,0 +1,235 @@
+            token = RefreshToken(value)
+        except Exception as e:
+            raise BadRequestException("유효하지 않은 리프레시 토큰입니다.", code="INVALID_REFRESH_TOKEN") from e
+
```
여기 아래 부분도 아마 토큰 검증할 떄 같이 수행될겁니다.

Meoyoug commented on src/users/oauth_serializers.py:
```diff
@@ -0,0 +1,235 @@
+        refresh = RefreshToken.for_user(user)
+        return str(refresh)
+
+    def _blacklist_existing_refresh_tokens(self, user):
```
이것도 직접 아웃스탠딩 토큰에 접근해서 리프레쉬 토큰 가져오기 보다 요청의 쿠키에 포함되어있는 리프레시 토큰을 가져와서 블랙리스트에 추가하는게 더 맞는 로직같아 보입니다.

SongMihyun replied:
아하 네엥~

Meoyoug commented on src/users/oauth_serializers.py:
```diff
@@ -0,0 +1,235 @@
+        outstanding_tokens.delete()
+
+
+class SocialLoginSerializer(serializers.Serializer):
```
요청을 통해 받은 리스폰스에서 이름, 이메일 , 폰넘버를 가져와서 유효성 검증을 진행한 후 유저 객체를 만드는 시리얼라이저 같은데 모델 시리얼라이저를 사용하지 않은 이유가 있을까요?

SongMihyun replied:
그런이유는 없지요 훗 // 의식의 흐름대로 그냥 생각나는대로 만들었기에.... 시정하도록 하지요 -0-

Meoyoug commented on src/users/oauth_serializers.py:
```diff
@@ -0,0 +1,235 @@
+            return requests.utils.unquote(clean_url)
+        return profile_image_url
+
+    def _blacklist_existing_refresh_tokens(self, user):
```
이것도 로그아웃 처리할 때만 해주면 될 것 같은데요.?

SongMihyun replied:
로그아웃 안되고 그냥 종료된경우에................... 빼도록 하지요...

Meoyoug commented on src/users/seriailzers.py:
```diff
@@ -1,17 +1,59 @@
         fields = ["id", "email", "name", "gender", "phone_number", "profile_image", "prefer_service", "prefer_location"]
         read_only_fields = ("id", "email")
 
+    def get_profile_image(self, obj):
```
이미지 URL을 임의로 바꿔서 내려주는 이유가뭔가요..?

SongMihyun replied:
/media/https%3A/phinf.pstatic.net/contact/68/2013/1/17/idblackcat_1358383585367.jpg

저장이과 나가는게 이런식으로 가길래요

Meoyoug reviewed (commented):
의문점이 정말 많습니다.

Commit:
feedback 수정

Commit:
test.sh

Meoyoug reviewed (approved):
고생하셨습니다
